README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
cms <img src="https://talbotsr.com/cms/logo.png" align="right" height="139" />
==============================================================================

Composite measure schemes (cms)
===============================

<!-- badges: start -->
[![Travis build status](https://travis-ci.org/mytalbot/cms.svg?branch=master)](https://travis-ci.org/r-lib/usethis) <!-- badges: end -->

The cms package uses tabular data from rat epilepsy studies and applies a composite measures scheme (via PCA) to select the most prominent features. Further, variables can be selected to perform cluster analysis on a subset in order to build a composite score. Finally, the cluster distribution is displayed for the subgroups and allows severity assessment between animal models.

Dependencies
------------

The cms package has some dependencies. We advise installing/updating the following packages before using cms:

-   ade4
-   scatterplot3d
-   lattice
-   corrplot
-   made4
-   caret
-   ggplot2
-   reshape2
-   plyr
-   RColorBrewer
-   gplots
-   e1072

Installation
------------

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mytalbot/cms")
library(cms)
```

Example
-------

The following example uses the (pre-cleaned) internalized epilepsy data (episet\_full) set with three experimental subgroups. Further, the feature selection is repeated 100-fold. The variable `scorevars=NA` ensures that the most prominent features in the subgroups are determined. For cluster distribution analysis specific variables must be selected in `scorevars` (see Vignette).

``` r
library(cms)
#> Loading required package: ade4
cms_cl        <- cms_clusters(episet_full, 
                              runs        = 100, 
                              emptysize   = 0.2, 
                              trainsize   = 0.8, 
                              idvariable  = "animal_id", 
                              varstart    = 14, 
                              exclude     = "Seizures_n",
                              scorevars   =  NA) 
cms_cl
#>                      x freq  perc
#> 1        burrowing_rat  100 25.00
#> 8   social_interaction  100 25.00
#> 6        openfield_rat   95 23.75
#> 7            Sacc_pref   54 13.50
#> 4 Of_immobile_duration   36  9.00
#> 5 OF_rearing_frequency    6  1.50
#> 3        BWB_streching    5  1.25
#> 2               BWB_LT    4  1.00
```


NEWS.md
# cms 0.0.9000

* Added a `NEWS.md` file to track changes to the package.


testscript.R
library(usethis)
library(devtools)

usethis::use_description()
package.skeleton(name = "cms")
use_r("cms_load")
use_r("cms_clusters")
use_r("cms_analysis")

install_github("mytalbot/cms")


library(cms)
file          <- "C:/MHH Bleich/Aktuelles/Maarten Paper/Maarten Epilepsy Score/all_raw.txt"
mydat         <- cms_load(file, remove_this="elec_tel_no SE")

cms_cl        <- cms_clusters(mydat, 
                              runs        = 100, 
                              emptysize   = 0.2, 
                              trainsize   = 0.8, 
                              idvariable  = "animal_id", 
                              varstart    = 14, 
                              exclude     = "Seizures_n",
                              
                              scorevars   =  c("Sacc_pref", "social_interaction", "burrowing_rat", "openfield_rat"))

thresholds           <- cms_cl[[1]] 
cluster_distribution <- cms_cl[[2]]  

cms_analysis(mydat, thresholds, cluster_distribution, savepath="C:/MHH Bleich/Packages/cms/cms")












R/cms_analysis.R
#' Composite measurement analysis function
#' 
#' The function takes in the results from the cms_clusters function (frequency distributions of each run of the 80 percent subsampling of the training data) and calculates
#' general attributes as well as graphical distribution outputs.
#'
#' @param raw requires the raw data list (also input in the cms_load function)
#' @param thresholds specification of the threshold values from the listed output of cms_clusters 
#' @param cluster_distribution output from the cms_cluster function (IMPORTANT: scorevars requires a vector of input variables!)
#' @param savepath include the path for the baplot here (no name spec for the plot necessary, it is called "Cluster_distros.png")
#' 
#' @import ggplot2
#' @import utils
#' @importFrom stats sd median
#' 
#' @return Data frame with the cluster distributions/severity attributions.+
#'
#' @export
#'
#'

cms_analysis <- function(raw, thresholds, cluster_distribution, savepath=NA){

  # calculate 95% CI for the thresholds between clusters ---------------------
  d          <- raw[raw[,"repeats"]==3,]
  
  mean(thresholds$threshold_1_2, na.rm = T) + (1.96*(sd(thresholds$threshold_1_2, na.rm = T)))/sqrt(NROW(d))
  mean(thresholds$threshold_1_2, na.rm = T) - (1.96*(sd(thresholds$threshold_1_2, na.rm = T)))/sqrt(NROW(d))
  
  mean(thresholds$threshold_2_3, na.rm = T) + (1.96*(sd(thresholds$threshold_2_3, na.rm = T)))/sqrt(NROW(d))
  mean(thresholds$threshold_2_3, na.rm = T) - (1.96*(sd(thresholds$threshold_2_3, na.rm = T)))/sqrt(NROW(d))
  
  
  
  # cluster distribution results --------------------------------------------
  # take the mean across all runs
  distr_means <- aaply(laply(cluster_distribution, as.matrix), c(2, 3), mean)
  # take the sd across all runs
  distr_sds   <- aaply(laply(cluster_distribution, as.matrix), c(2, 3), sd)
  
  # calculate percentage for each run 
  distr_perc  <- NULL
  distr_perc  <- lapply(cluster_distribution, function(x){
    apply(x, 1, function(y){100/sum(y)*y})
  })
  
  # mean sd over percentages 
  distr_perc_sd <- aaply(laply(distr_perc, as.matrix), c(2, 3), sd)
  
  # calculate in percentage 
  distr_means <- apply(distr_means, 1, function(x){100/sum(x)*x})
  
  # melt data for the graph
  distr_means <- melt(distr_means)
  
  #split the grouping name into two columns
  DM <- cbind(distr_means, 
                       data.frame(do.call(rbind, strsplit(as.vector(distr_means$X1), split = "_"))))
  
  #re label the column names
  distr_means$cluster   <- NULL
  distr_means$treat_mod <- NULL
  distr_means$perc      <- NULL
  distr_means$mod       <- NULL
  distr_means$treat     <- NULL
  
  distr_means$cluster   <- DM[,1]
  distr_means$treat_mod <- DM[,2]
  distr_means$perc      <- DM[,3]
  distr_means$mod       <- DM[,4]
  distr_means$treat     <- DM[,5]
  
  #names(distr_means)  <- c("cluster","treat_mod","perc","mod","treat")
 
  
  # change the names and order of the different groups (to make them appear in the graph the way we want)
  distr_means$treat       <- factor(distr_means$treat, levels=c("naive","sham","SE"))
  distr_means$mod         <- factor(distr_means$mod, levels=c("kind","chem","elec"))
  levels(distr_means$mod) <- c("Kindling","Chemical Post-SE","Electrical Post-SE")
  distr_means$cluster     <- factor(distr_means$cluster, levels=c("cluster3","cluster2","cluster1"))
  
  # standard deviation across groups in perc 
  distr_perc_sd[distr_perc_sd==0] <- NA
  apply(distr_perc_sd, 2, mean, na.rm=T)
  
  # create new column to get the group names right 
  # the "SE" group is split in "kindled" for the kindling model and "post-SE" for the other two models 
  distr_means$n_treat <- NULL
  distr_means$n_treat <- as.character(distr_means$treat)
  distr_means[distr_means [,"mod"] =="Kindling" & distr_means[,"treat"]=="SE","n_treat"] <- "Kindled"
  distr_means[!distr_means[,"mod"] =="Kindling" & distr_means[,"treat"]=="SE","n_treat"] <- "Post-SE"
  
  # re-name and re-order some things to make them appear nice in the graph
  distr_means$n_treat         <- as.factor(distr_means$n_treat)
  levels(distr_means$n_treat) <- c("Kindled","Naive","Post-SE","Sham")
  distr_means$n_treat         <- factor(distr_means$n_treat, levels=c("Naive","Sham","Kindled","Post-SE"))
  levels(distr_means$cluster) <- c("cluster1","cluster2","cluster3")
  distr_means$cluster         <- factor(distr_means$cluster, levels=c("cluster3","cluster2","cluster1"))
  
  #print(head(distr_means))
  
  # create graph of the mean distributions of the 100 runs 
  p <- ggplot(distr_means, aes(x=n_treat, y=perc, fill=cluster)) +
    geom_bar(stat = "identity") +
    facet_grid(~mod, scales="free") +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    scale_fill_brewer(palette="Reds", direction=-1) +
    labs(x="",y="distribution [%]")
  
  if(sum(is.na(savepath))==TRUE){ 
    
  }else{
    ggsave("Cluster_distros.png", plot = last_plot(), device = NULL, path = savepath,
           scale = 1, width = 6, height = 6, dpi = 300)
  }
    

  return(list(distr_means, p))
}




R/cms_clusters.R
#' Cluster and frequency distribution of composite measures for severity attribution
#'
#' The algorithm takes the cleaned input data, does some curation for missing values and then calculates the composite measures schemes.
#' The function can be used in two ways: 1. with the scorevars set to NA the data are screened for the most prominent feature contributions; 2. 
#' the so found feature sets can be specified, e.g., scorevars=c("Sacc_pref", "social_interaction", "burrowing_rat", "openfield_rat"). This will
#' result in the calculation of a severity-based cluster distribution analysis using the input variables.
#'
#' @param raw raw data (unprocessed or from cms_load)
#' @param runs number of random samplings for the PCA-based feature determination
#' @param emptysize fraction of empty data that will be removed
#' @param trainsize size of the "training data" in the fold subsamplings
#' @param idvariable column header of the animal id variable in the data
#' @param varstart column number with the first variable to start with - there shall be no general infomation on the right side of this number
#' @param exclude naming column headers that shall be excluded from the data set
#' @param scorevars if NA (default) the function will determine the most prominent variables in the sample data; NA can also be replaced by a vector of variables
#' 
#' @import made4  
#' @import ade4 
#' @import RColorBrewer 
#' @import gplots
#' @import scatterplot3d
#' @import lattice
#' @import caret
#' @import reshape2
#' @import plyr
#' @import corrplot
#' @importFrom stats kmeans predict   
#'
#' @return A list with composite score thresholds and k-means cluster attributions.
#' 
#' @export
#'
#'
cms_clusters  <- function(raw= raw, runs=100, emptysize=0.2, trainsize=0.8, idvariable="animal_id", varstart=14, exclude="Seizures_n",
                          scorevars=NA){

  # prepare the framework
  PC1loadings <- data.frame(matrix(nrow = 100, ncol = 10))
  PC2loadings <- data.frame(matrix(nrow = 100, ncol = 10))
  PCA_perc    <- NULL
  thresholds  <- data.frame("threshold_1_2"=NA,"threshold_2_3"=NA)
  cluster_distribution <- list()
  
  for(j in 1:runs){
    set.seed(j)
    
    # train and test set
    d        <- raw[raw[,"repeats"]==3,]
    d        <- d[, -which(colMeans(is.na(d)) > emptysize)]
    
    subset80 <- sample(d[[idvariable]], trainsize*length(d[[idvariable]]))
    d_80     <- d[ d[,idvariable] %in% subset80,]
    d_20     <- d[!d[,idvariable] %in% subset80,]
    
    # clean data and exclude single variables
    d_pca          <- d_80[,varstart:dim(d_80)[2]]
    d_pca[exclude] <- NULL
    
    # curate missing values in PCA data & transform data (BoxCox)
    missing <- c()
    for(i in 1:ncol(d_pca)){
      missing[i]                 <- sum(is.na(d_pca[,i]))
      d_pca[is.na(d_pca[,i]), i] <- mean(d_pca[,i], na.rm = TRUE)
    }
    #print(paste("Missing data: ", sum(missing),"(",round((sum(missing)/ (dim(d_pca)[1]*dim(d_pca)[2]))*100,2),"%)", sep="")) 
    
    pre.proc  <- preProcess(d_pca, method=c("BoxCox", "center", "scale"))
    d_pca     <- predict(pre.proc, d_pca)
    
    # do the PCA
    d_pca_ord <- ord((d_pca), type="pca")
    #print(head(d_pca))
    
    ### extract the LOADINGs and sort them
    temp            <- sqrt((d_pca_ord$ord$co[,1:2])^2)
    temp            <- temp[order(temp[,1], decreasing = T),]
    PC1loadings[j,] <- rownames(temp)[1:10]
    
    temp            <- temp[order(temp[,2], decreasing = T),]
    PC2loadings[j,] <- rownames(temp)[1:10]
    
    # save percentage explained
    PCA_perc        <- rbind(PCA_perc, (d_pca_ord$ord$eig * 100/sum(d_pca_ord$ord$eig))[1:5])
    
    # add PC1 and PC2 to the data sheet
    d_80$PC1        <- d_pca_ord$ord$li$Axis1
    d_80$PC2        <- d_pca_ord$ord$li$Axis2
    
    if(sum(is.na(scorevars))==TRUE){
      # nüscht
      
    }else{
      # Create a composite score based on n variables ---------------------------
      d_80$comp_score <- c(apply(scale(d_80[, scorevars]) ,1, sum, na.rm=T))
      d_80            <- d_80[!is.na(d_80$comp_score),]
      
      ### clustering of the score
      cl              <- kmeans(d_80$comp_score, 3)
      d_80$clusters   <- cl$cluster
      
      centers         <- data.frame(cluster=rownames(cl$centers), center=cl$centers)
      centers         <- centers[order(centers$center),] 
      centers$cluster_new_order <- 1:3
      
      
      # calculate the thresholds by checking the median between min and max of two clusters
      thresholds[j,]  <- data.frame("threshold_1_2"=
                                      median(c(  max(d_80[d_80[,"clusters"]==centers[1,"cluster"],"comp_score"]),
                                                 min(d_80[d_80[,"clusters"]==centers[2,"cluster"],"comp_score"])))
                                    ,"threshold_2_3"=
                                      median(c(  max(d_80[d_80[,"clusters"]==centers[2,"cluster"],"comp_score"]),
                                                 min(d_80[d_80[,"clusters"]==centers[3,"cluster"],"comp_score"]))) )
      
      # add re labeled clusters to data set
      # the lowest cluster is always 1, the highest 3
      d_80[d_80[,"clusters"]==centers[1,"cluster"],"n_cluster"] <- 1
      d_80[d_80[,"clusters"]==centers[2,"cluster"],"n_cluster"] <- 2
      d_80[d_80[,"clusters"]==centers[3,"cluster"],"n_cluster"] <- 3
      
      
      d_80$treat_mod <- as.factor(paste(d_80$mod,d_80$treatment, sep="_"))
      d_80$n_cluster <- as.factor(d_80$n_cluster)
      
      # count the number of clusters for each group
      temp          <- data.frame(melt(tapply(d_80$n_cluster, d_80$treat_mod, table)))
      
      # save the results as a data.frame (not the most elegant solution), which is then stored in a list
      cluster_distribution[[length(cluster_distribution)+1]] <- data.frame(row.names = levels(d_80$treat_mod),
                                                                           cluster1=melt(lapply(temp$value, "[[", 1))$value,
                                                                           cluster2=melt(lapply(temp$value, "[[", 2))$value,
                                                                           cluster3=melt(lapply(temp$value, "[[", 3))$value)
    }
  }
  
  
  # Outside the Luhp - show the best parameter selections when no selection was provided
  if(sum(is.na(scorevars))==TRUE){
    temp      <- count(c(PC2loadings$X1,PC2loadings$X2,PC2loadings$X3,PC2loadings$X4))
    temp$perc <- 100/sum(temp$freq)*temp$freq
    return(temp[order(-temp$perc),])
  }else{ 
    return(list(thresholds, cluster_distribution))
  }
  
}
